File: server/src/services/scrapers/flipkart/flipkart.scraper.js
const cheerio = require('cheerio');
const BrowserFactory = require('../../shared/browserFactory');
const ScraperBase = require('../../shared/scraperBase');
const logger = require('../../shared/logger');
const DelayManager = require('../../antiblock/delayManager');
const RateLimiter = require('../../antiblock/rateLimiter');
const RetryStrategy = require('../../antiblock/retryStrategy');
const {
  normalizeFlipkartTitle,
  parsePrice,
  parseRating,
  parseReviewCount,
  determineCategory
} = require('./flipkart.normalizer');

// NOTE: Browser lifecycle and stealth scripts are managed by BrowserFactory.
// The globalBrowser singleton is owned by the factory, not this scraper.

class FlipkartScraper extends ScraperBase {
  constructor() {
    super({ sourceName: 'Flipkart', timeout: 35000 });
    this.baseUrl = 'https://www.flipkart.com';
  }

  buildSearchUrl(query) {
    const encodedQuery = encodeURIComponent(query);
    return `${this.baseUrl}/search?q=${encodedQuery}&marketplace=FLIPKART&as-show=on&as=off`;
  }

  async search(query) {
    this.logger.info(`Starting Playwright search for: "${query}"`);
    let context = null;
    let page = null;

    try {
      const pageHTML = await RetryStrategy.execute(async () => {
        await RateLimiter.throttle('flipkart.com', 3000);

        // ── 1. Hardened context from factory (stealth + correct UA pre-applied) ──
        context = await BrowserFactory.createHardenedContext({
          extraHTTPHeaders: {
            'Referer': 'https://www.flipkart.com/'
          }
        });

        // ── 2. Page with resource blocking from factory (allow images for lazy-loading) ──
        const allowedTypes = ['script', 'xhr', 'fetch', 'websocket', 'eventsource', 'manifest', 'other', 'document', 'image'];
        page = await BrowserFactory.createPage(context, allowedTypes);

        // ── 3. Cookie session pre-warm ──
        // Flipkart requires a valid session cookie before serving search results.
        this.logger.debug('[Flipkart] Pre-warming session via homepage...');
        await page.goto(this.baseUrl, { waitUntil: 'domcontentloaded', timeout: 20000 });
        await page.keyboard.press('Escape').catch(() => {});
        await DelayManager.randomSleep(700, 1400);

        // ── 4. Navigate to search ──
        const url = this.buildSearchUrl(query);
        const response = await page.goto(url, { waitUntil: 'domcontentloaded', timeout: 30000 });

        if (response && response.status() === 403) {
          throw new Error('Blocked by Flipkart WAF (403). Session may be stale.');
        }

        // ── 5. Wait for product grid ──
        const productSelector = 'div[data-id], div._1AtVbE, div.cPHDOP';
        await page.waitForSelector(productSelector, { state: 'attached', timeout: 15000 }).catch(() => null);

        // Force a multi-step scroll to trigger lazy-loading IntersectionObservers
        await page.evaluate(() => window.scrollBy(0, 800)).catch(() => {});
        await DelayManager.randomSleep(500, 800);
        await page.evaluate(() => window.scrollBy(0, 800)).catch(() => {});
        await DelayManager.randomSleep(1000, 1500);

        return await page.content();

      }, {
        context: `Flipkart Playwright [${query}]`,
        maxRetries: 2,
        baseDelayMs: 4000,
        isPlaywright: true
      });

      // ── 9. Parse with Cheerio ──
      const rawProducts = await this.extractData(pageHTML);
      if (!rawProducts || rawProducts.length === 0) {
        this.logger.warn(`No products parsed for: "${query}"`);
        return [];
      }

      const normalizedProducts = rawProducts.map(item => {
        try { return this.normalizeProduct(item); } catch { return null; }
      }).filter(Boolean);

      this.logger.info(`Successfully extracted ${normalizedProducts.length} products for: "${query}"`);
      return normalizedProducts;

    } catch (error) {
      this.logger.error(`Search failed for "${query}": ${error.message}`);
      return [];
    } finally {
      // Always clean up to prevent memory leaks in long-running server processes
      if (page) await page.close().catch(() => {});
      if (context) await context.close().catch(() => {});
    }
  }

  /**
   * Parses the Flipkart search result HTML using Cheerio.
   * data-id is a structurally stable anchor — Flipkart has never removed it.
   * All CSS class names are multi-fallback to survive Flipkart's weekly rotations.
   */
  async extractData(html) {
    const $ = cheerio.load(html);
    const results = [];

    // Primary anchor: data-id is the only structurally stable marker Flipkart has maintained.
    // Fallback anchors cover alternate grid layouts (cPHDOP = 2024 grid wrapper, _1AtVbE = older grid).
    const cardSelector = 'div[data-id], div.cPHDOP div[class*="col"], div._1AtVbE';

    $(cardSelector).each((i, el) => {
      const $el = $(el);

      // ── Title ──
      // Flipkart rotates CSS class names every 2-4 weeks. We cascade through 10 known variants.
      // Final fallback: img alt text (Flipkart always populates this with the product name).
      const rawTitle = $el.find([
        'div._4rR01T',    // Classic list layout (2021–2024)
        'div.KzDlHZ',    // Alternate list layout
        'a.IRpwTa',      // Category-specific link variant
        'a.s1Q9rs',      // Fashion grid
        'a.WKTcLC',      // Electronics list
        'div.wjcEIp',    // 2024 grid title
        'div._2WkVRV',   // 2025 grid title
        'a[class*="title"]',     // Generic class-fragment fallback
        'div[class*="title"]',
        'span[class*="title"]'
      ].join(', ')).first().text().trim() || $el.find('img').first().attr('alt')?.trim() || '';

      if (!rawTitle || rawTitle.length < 3) return;

      // ── Price ──
      // Attempt 1: Known class name variants
      let priceText = $el.find([
        'div._30jeq3',       // Classic price (2021–2024)
        'div.Nx9bqj',        // 2023 grid variant
        'div.hl05eU',        // 2024 variant
        'div._25b18c span',  // Nested price span
        'div.yRaY8j',        // 2025 variant
        'div[class*="price"]',
        'span[class*="price"]'
      ].join(', ')).first().text().trim();

      // Attempt 2: Look for any text node containing the ₹ rupee symbol
      if (!priceText) {
        $el.find('*').each((_, node) => {
          const text = $(node).clone().children().remove().end().text().trim();
          if (text.includes('₹') && text.length < 15) {
            priceText = text;
            return false; // break
          }
        });
      }

      if (!priceText) return;

      // ── Image ──
      let image = '';
      $el.find('img').each((_, img) => {
        const src = $(img).attr('src') || '';
        const lazySrc = $(img).attr('data-src') || $(img).attr('data-lazy-src') || '';
        const candidate = lazySrc || src;
        
        if (candidate && !candidate.includes('placeholder') && !candidate.includes('fa_') && !candidate.startsWith('data:image')) {
          image = candidate;
          return false; // break loop
        }
      });
      
      if (!image) {
        // Fallback: If no valid image found, it might be in a background-image div
        image = $el.find('[style*="background-image"]').first().css('background-image')?.replace(/url\(['"]?(.*?)['"]?\)/i, '$1') || '';
      }

      // ── Rating & Reviews ──
      const ratingText = $el.find([
        'div._3LWZlK', 'div.XQDdHH', 'div.ipqd2A',
        'span[class*="rating"]', 'div[class*="rating"]'
      ].join(', ')).first().text().trim();

      const reviewText = $el.find([
        'span._2_R_DZ', 'span.Wphh3N', 'span._13vcmD',
        'span[class*="review"]', 'span[class*="rating-count"]'
      ].join(', ')).first().text().trim();

      // ── Availability ──
      const outOfStock = $el.find('[class*="out-of-stock"], [class*="outOfStock"], [class*="sold-out"]').length > 0;
      const availability = outOfStock ? 'Out of Stock' : 'In Stock';

      // ── Link ──
      let link = $el.find('a[href*="/p/"]').first().attr('href') || $el.find('a').first().attr('href') || '';
      if (link && !link.startsWith('http')) link = this.baseUrl + link;
      // Strip tracking params but keep the /p/ product path
      if (link) link = link.split('?')[0];

      results.push({ rawTitle, priceText, image, ratingText, reviewText, link, availability });
    });

    this.logger.debug(`[Flipkart] Parsed ${results.length} raw items from HTML`);
    // PERF: Explicitly destroy Cheerio's parse5 DOM tree after extraction.
    // Without this, closures inside .each() callbacks keep the entire parsed HTML
    // tree referenced in heap, causing +30-50MB per scrape call that never GC'd.
    $.root().empty();
    return results;
  }

  normalizeProduct(rawProduct) {
    const title = rawProduct.rawTitle;
    const price = parsePrice(rawProduct.priceText);
    if (!title || price <= 0) {
      throw new Error(`Invalid data: title="${title}", price="${rawProduct.priceText}"`);
    }
    return {
      title,
      normalizedTitle: normalizeFlipkartTitle(title),
      price,
      currency: 'INR',
      image: rawProduct.image || '',
      rating: parseRating(rawProduct.ratingText),
      reviewCount: parseReviewCount(rawProduct.reviewText),
      source: this.sourceName,
      category: determineCategory(title),
      availability: rawProduct.availability || 'In Stock',
      link: rawProduct.link || '',
      timestamp: new Date().toISOString()
    };
  }
}

module.exports = FlipkartScraper;

File: ml-service/scrapers/flipkart_scraper.py
import logging
import urllib.parse
from bs4 import BeautifulSoup
from .base_scraper import BaseScraper

logger = logging.getLogger(__name__)

class FlipkartScraper(BaseScraper):
    def __init__(self, proxy=None):
        super().__init__(base_url="https://www.flipkart.com", proxy=proxy)

    def _extract_price(self, text):
        """Extract numeric price from text like '₹14,999'"""
        if not text: return 0
        cleaned = "".join(c for c in text if c.isdigit() or c == '.')
        try:
            return float(cleaned) if cleaned else 0
        except ValueError:
            return 0

    async def search(self, query, pages=2):
        """
        Search Flipkart using Playwright to render the page, 
        and BeautifulSoup to parse the extracted HTML.
        """
        results = []
        encoded_query = urllib.parse.quote_plus(query)
        
        for page_num in range(1, pages + 1):
            url = f"{self.base_url}/search?q={encoded_query}&page={page_num}"
            logger.info(f"Scraping Flipkart page {page_num}: {url}")
            
            html = await self.fetch_page_html(url)
            if not html:
                logger.warning(f"Failed to fetch HTML for page {page_num}")
                continue
                
            soup = BeautifulSoup(html, 'lxml')
            
            # Check for Captcha even with Playwright
            if "captcha" in soup.text.lower() or "verify you are human" in soup.text.lower():
                logger.error(f"Playwright hit a CAPTCHA at page {page_num}!")
                break # Stop paginating if we hit a captcha
            
            # Flipkart has different class structures depending on the product category.
            # Containers: _1AtVbE, _75nlfW (new), cPHDOP (newest)
            items = soup.find_all('div', attrs={'class': lambda e: e and ('_75nlfW' in e or 'cPHDOP' in e or '_1AtVbE' in e)})
            
            page_results = []
            for item in items:
                link_tag = item.find('a', class_=lambda c: c and ('CGtC98' in c or 'VJA3hP' in c or '_1fQZEK' in c or 'IRpwTa' in c or 'WKTcLC' in c))
                if not link_tag: continue
                    
                link = self.base_url + link_tag.get('href', '')
                
                title_tag = item.find('div', class_=lambda c: c and 'KzDlHZ' in c)
                if not title_tag:
                    title_tag = item.find('a', class_=lambda c: c and ('IRpwTa' in c or 'WKTcLC' in c))
                
                title = title_tag.text.strip() if title_tag else ""
                if not title or len(title) < 3: continue
                    
                brand_tag = item.find('div', class_=lambda c: c and 'syl9yP' in c)
                brand = brand_tag.text.strip() if brand_tag else ""
                if brand and not title.lower().startswith(brand.lower()):
                    title = f"{brand} {title}"

                price_tag = item.find('div', class_=lambda c: c and ('Nx9bqj' in c or '_30jeq3' in c))
                price = self._extract_price(price_tag.text) if price_tag else 0
                if price == 0: continue

                original_price_tag = item.find('div', class_=lambda c: c and ('yRaY8j' in c or '_3I9_wc' in c))
                original_price = self._extract_price(original_price_tag.text) if original_price_tag else price

                discount = 0
                if original_price > price:
                    discount = round(((original_price - price) / original_price) * 100)

                img_tag = item.find('img', class_=lambda c: c and ('DByuf4' in c or '_396cs4' in c))
                image = img_tag.get('src', '') if img_tag else ""
                
                rating_tag = item.find('div', class_=lambda c: c and ('XQDdHH' in c or '_3LWZlK' in c))
                rating = 0.0
                if rating_tag:
                    try: rating = float(rating_tag.text.strip()[:3])
                    except ValueError: pass
                
                review_tag = item.find('span', class_=lambda c: c and ('Wphh3N' in c or '_2_R_DZ' in c))
                review_count = 0
                if review_tag:
                    txt = review_tag.text.strip().split(' ')[0]
                    review_count = int(self._extract_price(txt))
                    
                oos_tag = item.find('div', string=lambda s: s and 'Out of stock' in s)
                availability = "out_of_stock" if oos_tag else "in_stock"

                product_data = {
                    "platform": "flipkart",
                    "title": title,
                    "brand": brand,
                    "price": price,
                    "originalPrice": original_price,
                    "discount": discount,
                    "currency": "INR",
                    "image": image,
                    "link": link,
                    "rating": rating,
                    "reviewCount": review_count,
                    "availability": availability
                }
                
                if not any(p['link'] == link for p in page_results):
                    page_results.append(product_data)
                    
            logger.info(f"Found {len(page_results)} items on page {page_num}")
            results.extend(page_results)
            if not page_results: break
                
        return results

# Expose async function
async def scrape_flipkart_async(query, pages=2):
    scraper = FlipkartScraper()
    return await scraper.search(query, pages=pages)
